### ISSUE TITLE
BCDB Index destroyed when redis restarted [development_installer]

### ISSUE THREAD

**Issue by arahmanhamdy** [threefoldtecharchive/jumpscaleX_archived#559]
BCDB indexes the records into redis. The issue is that with restarting redis it deletes all these indeces. We need to have a consistent way for creating indexes.

**Comment by despiegk** [comment]
for now we rely on redis to no stop.
there is an export/import feature now on BCDB which allows a developer to maintain state and still restart redis.

there is also an index_rebuild which can be used

**Comment by despiegk** [comment]
we can put something in that if the redis is empty we rebuild

**Comment by despiegk** [comment]
we should have a jsx bcdb_check , should be separate FR

**Comment by arahmanhamdy** [comment]
The issue is when/how to check for the index loading? Is there a way to check if there is no loaded indexes and reload for all models? Do we have a place where we can get all models?

**Comment by despiegk** [comment]
![image](https://user-images.githubusercontent.com/6021844/59384036-88531080-8d61-11e9-8c1f-8d728d876179.png)
fixed

**Comment by Pishoy** [comment]
- with etcd client, client is still exist after restart docker as below 
- with MySQL client, client doest not exist after restart docker as below, so there is issue exist with MySQL client 

- etcd client 
```
JSX> j.clients.etcd.get('etcd1')                                                                                                                                                                            
## j.clients.etcd Instance: 'etcd1' 

- host                          : 127.0.0.1
- name                          : etcd1
- password_                     : 
- port                          : 2379
- user                          : 
pishoy@pishoy-laptop:~$ docker restart d6c87d00148d
d6c87d00148d
pishoy@pishoy-laptop:~$ ssh root@localhost -p9122
OK
3BOTDEVEL:3bot:~: kosmos 
MYENV INIT
MYENV INIT

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
redis-server is already the newest version (5:4.0.9-1ubuntu0.1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
sysctl: setting key "vm.overcommit_memory": Read-only file system
Thu 13 16:06:34 InstallTools.py  - 157 - core_get                           : mkdir -p /sandbox/var;redis-server --unixsocket /sandbox/var/redis.sock --port 6379 --maxmemory 100000000 --daemonize yes
50:C 13 Jun 16:06:34.714 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
50:C 13 Jun 16:06:34.714 # Redis version=4.0.9, bits=64, commit=00000000, modified=0, pid=50, just started
50:C 13 Jun 16:06:34.714 # Configuration loaded
Thu 13 16:06:34 BCDB.py          - 337 - bcdb:bcdb                          : REBUILD INDEX
Thu 13 16:06:34 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.bcdb.acl.circle.1
Thu 13 16:06:34 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:06:34 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:06:34 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:06:34 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.bcdb.acl.circle.1

Thu 13 16:06:34 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.bcdb.acl.user.1
Thu 13 16:06:34 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:06:34 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:06:34 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:06:34 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.bcdb.acl.user.1

Thu 13 16:06:34 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:06:34 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:06:34 BCDBModel.py     - 187 - acl:acl                            : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:06:34 BCDBModel.py     - 199 - acl:acl                            : will rebuild index for:model:jumpscale.bcdb.acl.1

Thu 13 16:06:34 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:06:34 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:06:34 BCDBModel.py     - 187 - user:user                          : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:06:34 BCDBModel.py     - 199 - user:user                          : will rebuild index for:model:jumpscale.bcdb.user.1

Thu 13 16:06:34 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:06:34 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:06:34 BCDBModel.py     - 187 - circle:circle                      : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:06:34 BCDBModel.py     - 199 - circle:circle                      : will rebuild index for:model:jumpscale.bcdb.circle.1

Thu 13 16:06:34 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:06:34 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:06:34 BCDBModel.py     - 187 - namespace:namespace                : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:06:34 BCDBModel.py     - 199 - namespace:namespace                : will rebuild index for:model:jumpscale.bcdb.namespace.1

Thu 13 16:06:34 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.mysql.client
Thu 13 16:06:34 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:06:34 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:06:34 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:06:34 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.mysql.client

Thu 13 16:06:34 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.itsyouonline.jwt.1
Thu 13 16:06:34 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:06:34 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:06:34 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:06:34 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.itsyouonline.jwt.1

Thu 13 16:06:34 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.itsyouonline.1
Thu 13 16:06:34 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:06:34 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:06:34 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:06:34 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.itsyouonline.1

Thu 13 16:06:34 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.etcd.client
Thu 13 16:06:34 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:06:34 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:06:34 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:06:34 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.etcd.client

Thu 13 16:06:34 BCDB.py          - 230 - bcdb:bcdb                          : BCDB INIT DONE:system
JSX> j.clients.etcd.get('etcd1')                                                                                                                                                                            
## j.clients.etcd Instance: 'etcd1' 

- host                          : 127.0.0.1
- name                          : etcd1
- password_                     : 
- port                          : 2379
- user                          : 
```
- MySQL client
```
JSX> j.clients.mysql.get(name="test_sql",ipaddr="127.0.0.1",port=3306,login="newuser",passwd="password",dbname="testdb")                                                                                    
## j.clients.mysql Instance: 'test_sql' 

- dbname                        : testdb
- ipaddr                        : 127.0.0.1
- login                         : newuser
- name                          : test_sql
- passwd                        : password
- port                          : 3306





JSX> j.clients.mysql.find()                                                                                                                                                                                 
[]

JSX> j.clients.mysql.test_sql                                                                                                                                                                               
## j.clients.mysql Instance: 'test_sql' 

- dbname                        : testdb
- ipaddr                        : 127.0.0.1
- login                         : newuser
- name                          : test_sql
- passwd                        : password
- port                          : 3306

pishoy@pishoy-laptop:~$ docker restart d6c87d00148d
d6c87d00148d
pishoy@pishoy-laptop:~$ ssh root@localhost -p9122
OK
3BOTDEVEL:3bot:~: kosmos 
MYENV INIT
MYENV INIT

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
redis-server is already the newest version (5:4.0.9-1ubuntu0.1).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
sysctl: setting key "vm.overcommit_memory": Read-only file system
Thu 13 16:21:18 InstallTools.py  - 157 - core_get                           : mkdir -p /sandbox/var;redis-server --unixsocket /sandbox/var/redis.sock --port 6379 --maxmemory 100000000 --daemonize yes
50:C 13 Jun 16:21:18.204 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
50:C 13 Jun 16:21:18.204 # Redis version=4.0.9, bits=64, commit=00000000, modified=0, pid=50, just started
50:C 13 Jun 16:21:18.204 # Configuration loaded
Thu 13 16:21:18 BCDB.py          - 337 - bcdb:bcdb                          : REBUILD INDEX
Thu 13 16:21:18 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.bcdb.acl.circle.1
Thu 13 16:21:18 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:21:18 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:21:18 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:21:18 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.bcdb.acl.circle.1

Thu 13 16:21:18 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.bcdb.acl.user.1
Thu 13 16:21:18 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:21:18 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:21:18 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:21:18 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.bcdb.acl.user.1

Thu 13 16:21:18 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:21:18 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:21:18 BCDBModel.py     - 187 - acl:acl                            : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:21:18 BCDBModel.py     - 199 - acl:acl                            : will rebuild index for:model:jumpscale.bcdb.acl.1

Thu 13 16:21:18 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:21:18 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:21:18 BCDBModel.py     - 187 - user:user                          : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:21:18 BCDBModel.py     - 199 - user:user                          : will rebuild index for:model:jumpscale.bcdb.user.1

Thu 13 16:21:18 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:21:18 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:21:18 BCDBModel.py     - 187 - circle:circle                      : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:21:18 BCDBModel.py     - 199 - circle:circle                      : will rebuild index for:model:jumpscale.bcdb.circle.1

Thu 13 16:21:18 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:21:18 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:21:18 BCDBModel.py     - 187 - namespace:namespace                : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:21:18 BCDBModel.py     - 199 - namespace:namespace                : will rebuild index for:model:jumpscale.bcdb.namespace.1

Thu 13 16:21:18 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.mysql.client
Thu 13 16:21:18 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:21:18 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:21:18 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:21:18 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.mysql.client

Thu 13 16:21:18 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.itsyouonline.jwt.1
Thu 13 16:21:18 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:21:18 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:21:18 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:21:18 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.itsyouonline.jwt.1

Thu 13 16:21:18 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.itsyouonline.1
Thu 13 16:21:18 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:21:18 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:21:18 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:21:18 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.itsyouonline.1

Thu 13 16:21:18 BCDB.py          - 452 - bcdb:bcdb                          : load model:jumpscale.etcd.client
Thu 13 16:21:18 BCDB.py          - 240 - bcdb:bcdb                          : DATAPROCESSOR STARTS
Thu 13 16:21:18 BCDB.py          - 252 - bcdb:bcdb                          : DATAPROCESSOR STOPS
Thu 13 16:21:18 BCDBModel.py     - 187 -                                    : DATAPROCESSOR & SQLITE STOPPED OK
Thu 13 16:21:18 BCDBModel.py     - 199 -                                    : will rebuild index for:model:jumpscale.etcd.client

Thu 13 16:21:18 BCDB.py          - 230 - bcdb:bcdb                          : BCDB INIT DONE:system
JSX> j.clients.mysql.test_sql                                                                                                                                                                               
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/sandbox/lib/jumpscale/Jumpscale/core/BASECLASSES/JSBaseConfigs.py", line 211, in __getattr__
    r = self.get(name=name, die=False, create_new=False)
  File "/sandbox/lib/jumpscale/Jumpscale/clients/mysql/MySQLFactory.py", line 21, in get
    return self.getClient(name=name, id=id, die=die, create_new=create_new, **kwargs)
  File "/sandbox/lib/jumpscale/Jumpscale/clients/mysql/MySQLFactory.py", line 26, in getClient
    key = "%s_%s_%s_%s_%s" % (cl.ipaddr, cl.port, cl.login, cl.passwd, cl.dbname)
AttributeError: 'NoneType' object has no attribute 'ipaddr'

'NoneType' object has no attribute 'ipaddr'
JSX> j.clients.mysql.find                                                                                                                                                                                   
##  j.clients.mysql MySQLFactory

JSX> j.clients.mysql.find()                                                                                                                                                                                 
[]

```
- JSX 
```
3BOTDEVEL:3bot:~: cd /sandbox/code/github/threefoldtech/jumpscaleX/
3BOTDEVEL:3bot:jumpscaleX: git branch
  development
* development_installer
  development_jumpscale
3BOTDEVEL:3bot:jumpscaleX: git log -1
commit 914dce4833db599b16fcd3b6be1202b6a56bad6e (HEAD -> development_installer, origin/development_installer)
Merge: 51ae1334 2277399c
Author: kristof de spiegeleer <>
Date:   Thu Jun 13 16:37:37 2019 +0200

    Merge branch 'development' into development_installer
    
    # Conflicts:
    #       Jumpscale/builder/apps/BuilderAppsFactory.py
    #       Jumpscale/jumpscale_generated.py
```
```

**Comment by despiegk** [comment]
this bug is too hard to follow, what has ETCD to do with this?
will close this one

the development_baseclasses is resolving lots of issues lets restart testing and see where we get 
a new bug will be opened if we find something
